### PR TITLE
Feature/16 19/profile login screens

### PR DESCRIPTION
**Pull request by brandonlee503**
MVP version

### CODE REVIEW

File: C7FIT/AbstractHealthCell.swift [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    
+    public override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+        picker.delegate = self
+        picker.dataSource = self
+        setup()
+        setupConstraints()
+    }
+    
+    public required init?(coder aDecoder: NSCoder) {
```
Comment (splons) [inline C7FIT/AbstractHealthCell.swift:35]: really is a pain to have to litter the code with this weird required init

File: C7FIT/AbstractHealthCell.swift [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    open override var canBecomeFirstResponder: Bool {
+        return true
+    }
+    
+    open override var canResignFirstResponder: Bool {
+        return true
+    }
+    
+    open override func becomeFirstResponder() -> Bool {
+        picker.dataSource = self
```
Comment (splons) [inline C7FIT/AbstractHealthCell.swift:50]: is there a reason to reset this here? gets initially set in init above

File: C7FIT/AbstractHealthCell.swift [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    open func setup() {
+        dataTitle.backgroundColor = .red
+        addSubview(dataTitle)
+        
+        dataLabel.backgroundColor = .cyan
+        addSubview(dataLabel)
+    }
+    
+    open func setupConstraints() {
+        dataTitle.translatesAutoresizingMaskIntoConstraints = false
```
Comment (splons) [inline C7FIT/AbstractHealthCell.swift:75]: Optionally you can just wrap the array from the activate call around these constraints and then you don't have to create a let constant for each of them.

```
NSLayoutConstraint.activate([
	dataTitle.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10),
 	labelTop = dataTitle.topAnchor.constraint(equalTo: topAnchor, constant: 10),
 	labelBottom = dataTitle.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -10)
 ])
```

File: C7FIT/AbstractHealthCell.swift [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        let labelLeading = dataTitle.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10)
+        let labelTop = dataTitle.topAnchor.constraint(equalTo: topAnchor, constant: 10)
+        let labelBottom = dataTitle.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -10)
+        NSLayoutConstraint.activate([labelLeading, labelTop, labelBottom])
+        
+        dataLabel.translatesAutoresizingMaskIntoConstraints = false
```
Comment (splons) [inline C7FIT/AbstractHealthCell.swift:81]: I have migrated to using the anchor-based constraints for the most part too. Almost all the goodness of pure layout without the 3rd party sources.

File: C7FIT/FirebaseDataManager.swift [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    func updateUser(uid: String, user: User) {
+        let newUserRef = self.ref.child("users").child(uid)
+        newUserRef.setValue(user.toAnyObject())
+    }
+    
+    /**
+        Uploads new profile picture to Firebase Storage
+     
+        - Parameter uid: User's universal ID
+        - Parameter user: User data to update
```
Comment (splons) [inline C7FIT/FirebaseDataManager.swift:118]: the user param is copy/paste error from above?

File: C7FIT/FirebaseDataManager.swift [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    func uploadProfilePicture(uid: String, data: Data, completion: @escaping (_: URL?) -> Void) {
+        let storageRef = FIRStorage.storage().reference(withPath: "profilePics/\(uid).jpg")
+        let uploadMetaData = FIRStorageMetadata()
+        uploadMetaData.contentType = "image/jpeg"
+        storageRef.put(data, metadata: uploadMetaData) { (metaData, error) in
+            if error == nil {
+                // Update user profilePicURL
+                completion(metaData?.downloadURL())
+            } else {
+                print("Upload profile pic error: \(error?.localizedDescription)")
```
Comment (splons) [inline C7FIT/FirebaseDataManager.swift:131]: do you need to call completion(nil) here?

File: C7FIT/ProfileViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            let bio = json["bio"] as? String,
+            let weight = json["weight"] as? String,
+            let height = json["height"] as? String,
+            let bmi = json["bmi"] as? String,
+            let mileTime = json["mileTime"] as? String,
+            let pushups = json["pushups"] as? String,
+            let situps = json["situps"] as? String,
+            let legPress = json["legPress"] as? String,
+            let benchPress = json["benchPress"] as? String,
+            let lateralPull = json["lateralPull"] as? String else { return nil }
```
Comment (splons) [inline C7FIT/ProfileViewModel.swift:57]: minor point: would be nice to the return on its own line/block for visibility

File: C7FIT/User.swift [hunk 8]
```diff
@@ -13,24 +13,51 @@ import Firebase
...
-//    let mileTime: String?
-
-    init(authData: FIRUser) {
-//        uid = authData.uid
-        email = authData.email!
-    }
+    let photoURL: String?
+    let name: String?
+    let bio: String?
+    let weight: String?
```
Comment (splons) [inline C7FIT/User.swift:20]: If your deployment target is iOS 10 you should consider using NSMeasurment for values that incorporate unit information. Really nice new API.

https://developer.apple.com/reference/foundation/nsmeasurement

**Review by splons — COMMENTED** [review]
Overall looks very good

File: C7FIT/AbstractHealthCell.swift (hunk 3 above)
Comment (brandonlee503) [reply C7FIT/AbstractHealthCell.swift:75]: How does that work? I tried replacing my code my let constants and `NSLayoutConstraint.activate()` with your sample code but Xcode is complaining about unresolved identifiers?

File: C7FIT/FirebaseDataManager.swift (hunk 5 above)
Comment (brandonlee503) [reply C7FIT/FirebaseDataManager.swift:118]: Fixed

File: C7FIT/FirebaseDataManager.swift (hunk 6 above)
Comment (brandonlee503) [reply C7FIT/FirebaseDataManager.swift:131]: Fixed

File: C7FIT/AbstractHealthCell.swift (hunk 2 above)
Comment (brandonlee503) [reply C7FIT/AbstractHealthCell.swift:50]: Fixed

File: C7FIT/ProfileViewModel.swift (hunk 7 above)
Comment (brandonlee503) [reply C7FIT/ProfileViewModel.swift:57]: Fixed

File: C7FIT/User.swift (hunk 8 above)
Comment (brandonlee503) [reply C7FIT/User.swift:20]: Ohh man I wish I knew about this earlier! I think I may stick to keeping the values as Strings for now though and focus on refactoring/refining the other components of this app. Will look into this in the future.

File: C7FIT/AbstractHealthCell.swift (hunk 3 above)
Comment (splons) [reply C7FIT/AbstractHealthCell.swift:75]: The assignments to labelTop and labelBottom may be messing things up.

**Pull request merged**